server/context.ts

```diff
@@ -91,14 +100,34 @@ async function getAndUpdateSession(
   if (Object.keys(sessionUpdates).length > 0) {
     console.log("Updating session:", sessionUpdates);
 
-    prisma.session
-      .update({
-        where: { id: sessionId },
-        data: sessionUpdates,
-      })
-      .catch((error) => {
-        console.error("Error updating session:", error);
+    // Use the secure database function with SECURITY DEFINER
+    try {
+      // PgBouncer in transaction mode requires special handling
+      const authPrisma = new PrismaClient({
+        datasources: {
+          db: {
+            url: process.env.POSTGRES_PRISMA_URL,
+          },
+        },
       });
+      
+      // Call the upsert_session function using a simpler query format
+      // This works better with PgBouncer
+      await authPrisma.$executeRawUnsafe(`
```

Comment:
as we will be doing it for every requests we don't need to await for running it.